core/arch/arm/sm/suspend.S

```diff
@@ -0,0 +1,229 @@
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <kernel/unwind.h>
+#include <asm.S>
+#include <asm-defines.h>
+#include <arm.h>
+#include <arm32_macros.S>
+#include <arm32_macros_cortex_a9.S>
+#include <kernel/asan.h>
+#include <platform_config.h>
+#include <keep.h>
+#include <kernel/unwind.h>
+
+.section .text
+/*
+ * int cpu_suspend(uint32_t arg, int (*fn)(uint32_t))
+ * @arg will be passed to fn as argument
+ * return value: 0 - cpu resumed from suspended state.
+ *		 -1 - cpu not suspended.
+ */
+FUNC cpu_suspend, :
+UNWIND( .fnstart)
+	push	{r4 - r12, lr}
+UNWIND(.save	{r4 - r12, lr})
+	mov	r5, sp
+	sub	sp, sp, #SM_PM_CTX_SIZE
+	push	{r0, r1}
+
+	mov	r2, r5
+	mov	r1, #SM_PM_CTX_SIZE
+	add	r0, sp, #8
+	blx	cpu_suspend_save
+	adr	lr, aborted
+	/* Jump to arch specific suspend*/
+	pop	{r0, pc}
+aborted:
+	/* cpu not suspended */
+	add	sp, sp, #SM_PM_CTX_SIZE
+	/* Return -1 to the caller */
+	mov	r0, #(-1)
+	pop	{r4 - r12, pc}
+UNWIND( .fnend)
+END_FUNC cpu_suspend
+
+LOCAL_FUNC cpu_v7_suspend, :
+UNWIND(.fnstart)
+	push	{r4 - r11}
+	read_fcseidr r4
+	read_tpidruro r5
+	stmia	r0!, {r4 - r5}
+	read_dacr r4
+#ifdef CFG_WITH_LPAE
+#error "Not supported"
+#else
+	read_ttbr0 r5
+	read_ttbr1 r6
+	read_ttbcr r7
+#endif
+	read_sctlr r8
+	read_actlr r9
+	read_cpacr r10
+	read_mvbar r11
+	stmia	r0!, {r4 - r11}
+	read_prrr r4
+	read_nmrr r5
+	read_vbar r6
+	read_nsacr r7
+	stmia	r0, {r4 - r7}
+	pop	{r4 - r11}
+	bx	lr
+UNWIND( .fnend)
+END_FUNC cpu_v7_suspend
+
+FUNC cpu_do_suspend, :
+UNWIND(.fnstart)
+	push	{r4, r5}
+	read_midr r4
+	ubfx	r5, r4, #4, #12
+	ldr     r4, =CORTEX_A7_PART_NUM
+	cmp	r5, r4
+	beq	a7_suspend
+	ldr     r4, =CORTEX_A9_PART_NUM
+	cmp	r5, r4
+	beq	a9_suspend
+	/* cpu not supported */
+	b	.
+	/* A9 needs PCR/DIAG */
+a9_suspend:
+	read_pcr r4
+	read_diag r5
+	stmia	r0!, {r4 - r5}
+a7_suspend:
+	pop	{r4, r5}
+	b	cpu_v7_suspend
+UNWIND( .fnend)
+END_FUNC cpu_do_suspend
+
+FUNC cpu_resume, :
+UNWIND(.fnstart)
+	cpsid	aif
+
+	/*
+	 * At this point, MMU is not enabled now.
+	 * 1. Get the runtime physical address of _suspend_sp
+	 * 2. Get the offset from _suspend_sp to &thread_core_local
+	 * 3. Get the runtime physical address of thread_core_local
+	 * Since moving towards non-linear mapping,
+	 * `ldr r0, =thread_core_local` is not used here.
+	 */
+	adr	r0, _suspend_sp
+	ldr	r2, [r0]
+	add	r0, r0, r2
+
+	/* Assmue max 4 cpus and 1 cluster */
```

Comment:
Since this is generic code, can't we try to support more cpus and more than 1 cluster from start? 4x2 seems pretty common these days.

Reply:
@jbech-linaro This patchset is to support armv7 which mostly only contains 4 cores. Regarding armv8 aarch32, I do not have platforms to test. I could use simliar method as accessing thread_core_local here to use get_core_pos.

Reply:
True, well I think this would serve as a good start and we can deal with that question later on. I'm not familiar with imx-specfics, but all in all I think this looks good, thanks!